README.md
# Office Status using for Raspberry Pi with Pimoroni Unicorn hat

## Description

I have worked from home for a long time.  My home office is right next to the family room.  To make things easier to let them know if I was in a meeting, I wanted a visual way to let everyone know.  I used a Raspberry Pi W and a Pimoroni Unicorn Hat, some python programming, and some velcro to stick it to the family room TV.  

The python application has a built in webserver and GUI interface to both set the status and see what the current status is.

Additionally, I also have an [Office Status Stream Deck plugin](https://github.com/mattboston/officestatus-streamdeck), which not only let's me set the status but also view the status.  Check it out if you have a Stream Deck.

Coming Soon... I also have a LoupeDeck and will be looking into creating a LoupeDeck plugin.

## Hardware

* Raspberry Pi W
* Pimoroni Unicorn Hat (PIM498) - https://www.digikey.com/en/products/detail/pimoroni-ltd/PIM168/6928300
* Note: Other Pimoroni Unicorn Hats may work.  If you have another model, let me know if it works.

## Installation

* Configure your Raspberry Pi so that you can connect over the network
* ssh into the Raspberry Pi
* sudo git clone https://github.com/mattboston/officestatus.git /opt/officestatus
* cd /opt/officestatus
* bash install.sh
* sudo systemctl start officestatus

## How To Use

1. By default you can go to http://{raspberry_pi_ip_address} to use the built in GUI.
2. You can interact with the built in API 
3. You can download the OfficeStatus StreamDeck Plugin

## API

* GET /api/status - will give you a JSON output of the current status (free, away, available, busy)
* POST /api/setStatus - If you post JSON in the following format '{"status": "busy"}' using one of the statuses (free, away, available, busy)

## TODO

* tbd

## License

Please see the LICENSE file for licensing details.

## Changelog

Please see [CHANGELOG.md](CHANGELOG.md).

## Contributing

1. Fork the project
2. Make your changes, including tests that exercise the code
3. Summarize your changes in [CHANGELOG.md](CHANGELOG.md).
4. Make a pull request

## Author

Matt Shields [@mattboston]

## Screenshots

![In Use](screenshots/in_use.jpeg?raw=true "In Use")

![Raspberry Pi W & Pimoroni Unicorn Hat](screenshots/RaspberryPiW-PimoroniUnicornHat.jpeg?raw=true "Raspberry Pi W & Pimoroni Unicorn Hat")

![GUI Free](screenshots/gui_free.png?raw=true "GUI Free")

![GUI Busy](screenshots/gui_busy.png?raw=true "GUI Busy")

![GUI Available](screenshots/gui_available.png?raw=true "GUI Available")

![GUI Away](screenshots/gui_away.png?raw=true "GUI Away")

CHANGELOG.md
# Changelog

## 2021-07-08

### Initial commit

- While I've been using this for years, this is the initial upload of this code.

officestatus.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# =============================================================================
# Name:         Office Status
# Created By:   Matt Shields (GitHub @mattboston)
# Created Date: 2022-07-08
# License:      GPL v3
# =============================================================================

# =============================================================================
# Imports
# =============================================================================
import json
import unicornhat as unicorn
import threading
from time import sleep
import os.path
from pathlib import Path
import logging
import datetime
from random import seed
from random import random

from flask import Flask, jsonify, make_response, request, redirect, url_for, send_from_directory, render_template, send_from_directory
from random import randint

# =============================================================================
# Variables
# =============================================================================
debug = False
state_file = '/opt/officestatus/officestatus.txt'
log_file = '/opt/officestatus/officestatus.log'

# =============================================================================
# App
# =============================================================================
if debug == True:
    debug_level = logging.DEBUG
else:
    debug_level = logging.INFO
logging.basicConfig(filename=log_file, level=debug_level)
logging.info('%s - Started OfficeStatus app' % datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S') )

blinkThread = None
globalRed = 0
globalGreen = 0
globalBlue = 0

#setup the unicorn hat
unicorn.set_layout(unicorn.AUTO)
unicorn.brightness(0.5)
#get the width and height of the hardware
width, height = unicorn.get_shape()

app = Flask(__name__)

seed(1)

@app.route('/favicon.ico')
def favicon():
    return send_from_directory(os.path.join(app.root_path, 'static'), 'favicon.ico', mimetype='image/vnd.microsoft.icon')

def setColor(r, g, b, brightness, speed) :
    global crntColors, globalBlue, globalGreen, globalRed
    globalRed = r
    globalGreen = g
    globalBlue = b

    if brightness != '' :
        unicorn.brightness(brightness)

    for y in range(height):
        for x in range(width):
            unicorn.set_pixel(x, y, r, g, b)
    unicorn.show()

    if speed != '' :
        sleep(speed)
        unicorn.clear()
        crntT = threading.currentThread()
        while getattr(crntT, "do_run", True) :
             for y in range(height):
                 for x in range(width):
                     unicorn.set_pixel(x, y, r, g, b)
        unicorn.show()
        sleep(speed)
        unicorn.clear()
        unicorn.show()
        sleep(speed)

def switchOff() :
    global blinkThread
    if blinkThread != None :
        blinkThread.do_run = False
    unicorn.clear()
    unicorn.show()
    unicorn.off()

def checkStateFile():
    if not os.path.exists(state_file):
        logging.info("File doesn't exist. Creating new state file.")
        with open(state_file, 'w') as f:
            f.write('free')

def setStatus(status):
    f = open(state_file, "r+")
    f.truncate()
    f.write(status)
    f.close()
    logging.info('Status set to %s' % status)

def getStatus():
    f = open(state_file, "r")
    status = f.readline().splitlines()[0]
    f.close()
    logging.info('Status read %s' % status)
    return status

# API Initialization
@app.route('/')
def root():
    status = getStatus()
    return render_template("index.html", status=status)

@app.route('/off', methods=['POST'])
def off():
    setStatus('free')
    switchOff()
    return redirect("/", code=302)

@app.route('/busy', methods=["POST"])
def busy():
    setStatus('busy')
    switchOff()
    blinkThread = threading.Thread(target=setColor, args=(255, 0, 0, '', ''))
    blinkThread.do_run = True
    blinkThread.start()
    return redirect("/", code=302)

@app.route('/available', methods=['POST'])
def available():
    setStatus('available')
    switchOff()
    blinkThread = threading.Thread(target=setColor, args=(0, 255, 0, '', ''))
    blinkThread.do_run = True
    blinkThread.start()
    return redirect("/", code=302)

@app.route('/away', methods=['POST'])
def away():
    setStatus('away')
    switchOff()
    blinkThread = threading.Thread(target=setColor, args=(0, 0, 255, '', ''))
    blinkThread.do_run = True
    blinkThread.start()
    return redirect("/", code=302)

@app.route('/api/setStatus', methods=['POST'])
def apiSetStatus():
    request_data = request.get_json()
    logging.debug(request_data)
    status = request_data['status']
    setStatus(status)
    switchOff()
    if status == 'busy':
        blinkThread = threading.Thread(target=setColor, args=(255, 0, 0, '', ''))
        blinkThread.do_run = True
        blinkThread.start()
    elif status == 'available':
        blinkThread = threading.Thread(target=setColor, args=(0, 255, 0, '', ''))
        blinkThread.do_run = True
        blinkThread.start()
    elif status == 'away':
        blinkThread = threading.Thread(target=setColor, args=(0, 0, 255, '', ''))
        blinkThread.do_run = True
        blinkThread.start()
    elif status == 'free':
        switchOff()
    return jsonify({ 'status': status })
    
@app.route('/api/status', methods=['GET'])
def apiGetStatus():
    status = getStatus()
    return jsonify({ 'status': status })

@app.errorhandler(404)
def not_found(error):
    return make_response(jsonify({'error': 'Not found'}), 404)

if __name__ == '__main__':
    checkStateFile()
    app.run(host='0.0.0.0', debug=debug, port=80)
